### PR TITLE
rbindlist support fill=TRUE with use.names=FALSE and use it in merge.R ToDo of #678

### LINKED ISSUE #5262: R/merge.R:                # TO DO: use set() here instead..
Towards #678

Use `set` in `all.y` branch instead of `cbind`

### PR DESCRIPTION
**Pull request by ben-schwen**
Closes #5262

### merge.R
- [x] use `set` in merge.R instead of `cbind`

### CODE REVIEW

File: R/merge.R [hunk 1]
```diff
@@ -82,8 +82,9 @@ merge.data.table = function(x, y, by = NULL, by.x = NULL, by.y = NULL, all = FAL
       othercolsx = setdiff(nm_x, by)
       if (length(othercolsx)) {
         tmp = rep.int(NA_integer_, length(missingyidx))
-        # TO DO: use set() here instead..
-        yy = cbind(yy, x[tmp, othercolsx, with = FALSE])
+        nx = c(names(yy), names(x[tmp, othercolsx, with = FALSE]))
+        nx = make.unique(nx)
+        set(yy, NULL, tail(nx, -ncol(yy)), x[tmp, othercolsx, with = FALSE])
```
Comment (MichaelChirico) [inline R/merge.R:87]: is there a noticeable memory advantage to looping over the RHS columns instead of creating the whole with=FALSE subset table?

it looks like the advantage is we can do the tmp subset once & apply it to all columns...

**Review by MichaelChirico — APPROVED** [review]

File: R/merge.R (hunk 1 above)
Comment (ben-schwen) [reply R/merge.R]: I guessed that saving this tmp object and doing indexing twice is preferred over using additional memory. Rethinking it again, I might have figured out what the ToDo meant initially where using set omits this object.

**Comment by ben-schwen** [comment]
@MichaelChirico whats the preferred process here: 1 PR per source file? Or 1 PR per todo? Or 1 megathread PR?

**Comment by MichaelChirico** [comment]
I think generally one PR per TODO makes the most sense, but if you see a cluster of TODOs in closely related code, do combine them. basically TODOs that are logically independent should get their own PRs

File: R/merge.R [hunk 2]
```diff
@@ -81,9 +81,9 @@ merge.data.table = function(x, y, by = NULL, by.x = NULL, by.y = NULL, all = FAL
       yy = y[missingyidx]
       othercolsx = setdiff(nm_x, by)
       if (length(othercolsx)) {
-        tmp = rep.int(NA_integer_, length(missingyidx))
-        # TO DO: use set() here instead..
-        yy = cbind(yy, x[tmp, othercolsx, with = FALSE])
+        nx = c(names(yy), paste0("V",seq_len(length(othercolsx))))
+        nx = make.unique(nx)
+        set(yy, NULL, tail(nx, -ncol(yy)), rep(list(NA), length(othercolsx)))
```
Comment (MichaelChirico) [inline R/merge.R:86]: does this need to be NA_integer_?

Comment (ben-schwen) [reply R/merge.R:86]: Nope, `NA` logical works fine here. `rbindlist` coerces types so `logical` columns of `yy`will be coerced to the types of `dt`. The reason behind the previous `tmp` used `NA_integer_` is that it was used for indexing the data.table and therefore was `integer`.

File: R/merge.R [hunk 3]
```diff
@@ -81,9 +81,8 @@ merge.data.table = function(x, y, by = NULL, by.x = NULL, by.y = NULL, all = FAL
       yy = y[missingyidx]
       othercolsx = setdiff(nm_x, by)
       if (length(othercolsx)) {
-        tmp = rep.int(NA_integer_, length(missingyidx))
-        # TO DO: use set() here instead..
-        yy = cbind(yy, x[tmp, othercolsx, with = FALSE])
+        nx = make.unique(c(names(yy), othercolsx))
+        set(yy, NULL, tail(nx, length(othercolsx)), rep(list(NA), length(othercolsx)))
```
Comment (jangorecki) [inline R/merge.R:85]: What if missingindex is 1:nrow(y) and potentially copy is not happening (or may not happening in future due to some regression). Would be good to have a test that ensure that. Otherwise we could ended up updating y table.

**Review by jangorecki — COMMENTED** [review]
Btw. I recall cbindlist in a PR has option for controlling copy behavior.

**Comment by jangorecki** [comment]
Uh, sorry for late feedback but I realized nrow(y)=1 can be useful to test against as well. I can imagine that there could be different handling of a copy when you subset all rows and the index is scalar at the same time. Kind of edge case but worth to have it as well.

**Comment by ben-schwen** [comment]
> Uh, sorry for late feedback but I realized nrow(y)=1 can be useful to test against as well. I can imagine that there could be different handling of a copy when you subset all rows and the index is scalar at the same time (based on DT[TRUE] doing shallow copy). Kind of edge case but worth to have it as well.

Good point. I added test cases to check for accidental shallow copying.

AFAIU current behavior of shallow copying of #3215 (and friends) leads only to problems when pre-copy cols are altered in the copy since these cols point to the same address? Since we only add cols here and don't alter other cols in `yy`, I guessed there shouldn't be a problem here.

File: inst/tests/tests.Rraw [hunk 4]
```diff
@@ -1875,6 +1877,16 @@ test(630.1, merge(DT1,DT2,all.x=TRUE), setkey(adt(merge(adf(DT1),adf(DT2),by="a"
 
 test(631, merge(DT1,DT2,all.y=TRUE), data.table(a=c(2,3,5),total.x=c(NA,1,1),total.y=c(5,1,2),key="a"))
 test(631.1, merge(DT1,DT2,all.y=TRUE), setkey(adt(merge(adf(DT1),adf(DT2),by="a",all.y=TRUE)),a))
+# ensure merge(x,y,all.y) does not alter input y
+# merge containing idx 1:nrow(y)
```
Comment (MichaelChirico) [inline inst/tests/tests.Rraw:1881]: this comment is a little unclear on its own -- ideally we can read the comment without any more context and know the point of the test. I think 'idx' refers to the implementation? best to be more explicit about what it means

**Review by MichaelChirico — APPROVED** [review]

**Comment by mattdowle** [comment]
I tried to get to the root cause of this block of code. It took me a while to realize that all it was doing was adding on the right number of NA columns to `yy` so it could be `rbind`-ed using `use.names=FALSE`. So I made `rbind` support `use.names=FALSE` together with `fill=TRUE` and that block goes away.
I also removed the long standing comment about issue #24 after looking at it and thinking that comment is unlikely to be useful in the future now.
Left the new tests in place and unchanged.
